File: native/Sources/CrucibleCore/Powder/PowderEvents.swift
// The four set-piece events: a meteor, a blast, a surge of water, and a deep freeze.
//
// Ported from web/src/sim/powder-events.ts.
//
// Those rules used to live inside the web version's canvas component, tangled up with sound
// calls and screen-shake state, where nothing could test them and nothing could compare them
// against this. Extracting them on that side is what made this port checkable, and the two are
// now compared cell for cell like everything else.

/// One of the four set-piece events.
public enum PowderEventID: String, CaseIterable, Sendable, Hashable, Codable {
    /// A ball of lava and fire dropped from above, detonating when it lands.
    case meteor
    /// One large explosion at the centre, then two flanking it.
    case blast
    /// A wall of water down the left side, already moving.
    case surge
    /// Everything chilled at once: liquids to ice, lava to stone.
    case freeze
}

/// A noise an event wants made. The engine does not make it.
public enum PowderEventSound: String, Sendable, Hashable, Codable {
    case meteor
    case explosion
}

/// One explosion, as an event schedules it.
public struct PowderEventBlast: Sendable, Hashable, Codable {
    public var x: Int
    public var y: Int
    public var radius: Int
    public var force: Double
    public var heat: Double

    public init(x: Int, y: Int, radius: Int, force: Double, heat: Double) {
        self.x = x
        self.y = y
        self.radius = radius
        self.force = force
        self.heat = heat
    }
}

/// The delayed second half of an event.
///
/// The delay is part of the event's definition rather than a detail of how it is presented: a
/// meteor that detonates the instant it appears does not read as a meteor. So it is described
/// here and the app runs the timer.
public struct PowderEventFollowUp: Sendable, Hashable, Codable {
    public var delaySeconds: Double
    public var blasts: [PowderEventBlast]
    public var shake: Double
    public var sound: PowderEventSound?
    public var soundIntensity: Double
}

/// What accompanied the immediate half of an event, and what is still to come.
///
/// Shaking the screen and making a noise are not the simulation's business — the same division
/// the engine already makes for explosions through ``PowderEngine/onBurst`` — so they are
/// reported for the caller to act on.
public struct PowderEventStart: Sendable, Hashable, Codable {
    public var shake: Double
    public var sound: PowderEventSound?
    public var soundIntensity: Double
    public var followUp: PowderEventFollowUp?
}

extension PowderEngine {
    /// Runs the immediate half of an event.
    ///
    /// - Returns: what should accompany it, including the delayed half where it has one.
    @discardableResult
    public func start(_ event: PowderEventID) -> PowderEventStart {
        switch event {
        case .meteor: return startMeteor()
        case .blast: return startBlast()
        case .surge: return startSurge()
        case .freeze: return startFreeze()
        }
    }

    /// Runs the delayed half of an event.
    public func finish(_ followUp: PowderEventFollowUp) {
        for blast in followUp.blasts {
            triggerExplosion(
                centerX: blast.x,
                centerY: blast.y,
                radius: blast.radius,
                shockwaveForce: blast.force,
                maxHeat: blast.heat
            )
        }
    }

    // MARK: Meteor

    private func startMeteor() -> PowderEventStart {
        let centerX = width / 2
        let radius = 10
        // Dropped at a fixed depth from the top rather than at the top edge, so the ball is
        // fully inside the world and falls as one mass.
        let centerY = 14

        for dy in -radius ... radius {
            for dx in -radius ... radius {
                // A disc, compared without a square root.
                if dx * dx + dy * dy > radius * radius { continue }
                let x = centerX + dx
                let y = centerY + dy
                // Checked *before* the draw below. A meteor in a narrow world must consume
                // exactly one random number per cell it can actually fill, or every later
                // decision in the world shifts — which the recorded comparison would catch as a
                // mismatched draw count.
                guard isValid(x, y) else { continue }
                setElement(x, y, rng.next() < 0.8 ? Element.lava : Element.fire, temp: 2800)
                velocityY[index(x, y)] = 18
            }
        }

        return PowderEventStart(
            shake: 16,
            sound: .meteor,
            soundIntensity: 1,
            followUp: PowderEventFollowUp(
                delaySeconds: 0.18,
                // Placed at a fixed fraction of the world's height rather than wherever the
                // material actually reached, because it may not reach anywhere — it can land on
                // a ceiling someone built. Fixing it up front means the event always reads the
                // same.
                blasts: [
                    PowderEventBlast(
                        x: centerX,
                        y: Int((Double(height) * 0.65).rounded(.down)),
                        radius: 28,
                        force: 18,
                        heat: 3000
                    )
                ],
                shake: 22,
                sound: .explosion,
                soundIntensity: 2
            )
        )
    }

    // MARK: Blast

    private func startBlast() -> PowderEventStart {
        let centerX = width / 2
        let centerY = height / 2
        triggerExplosion(
            centerX: centerX,
            centerY: centerY,
            radius: 36,
            shockwaveForce: 22,
            maxHeat: 3500
        )

        return PowderEventStart(
            shake: 24,
            sound: .explosion,
            soundIntensity: 3,
            followUp: PowderEventFollowUp(
                delaySeconds: 0.12,
                blasts: [
                    PowderEventBlast(x: centerX - 22, y: centerY - 14, radius: 22, force: 16, heat: 2800),
                    PowderEventBlast(x: centerX + 22, y: centerY + 14, radius: 22, force: 16, heat: 2800),
                ],
                shake: 16,
                sound: nil,
                soundIntensity: 1
            )
        )
    }

    // MARK: Surge

    private func startSurge() -> PowderEventStart {
        let startY = Int((Double(height) * 0.28).rounded(.down))
        let endX = min(28, width - 4)
        let endY = height - 2

        // Written as a pair of while loops rather than over ranges. In a world too small or too
        // narrow for the band, the bounds cross over — and a Swift range whose end is below its
        // start is a crash, where JavaScript's loop simply does not run. A surge in a thirty-cell
        // world should do nothing, not bring the app down.
        var y = startY
        while y < endY {
            var x = 2
            while x < endX {
                setElement(x, y, Element.water, temp: 12)
                velocityX[index(x, y)] = 14
                x += 1
            }
            y += 1
        }

        return PowderEventStart(shake: 0, sound: nil, soundIntensity: 1, followUp: nil)
    }

    // MARK: Freeze

    private func startFreeze() -> PowderEventStart {
        // Writes the grid directly rather than going through `setElement`, which is deliberate
        // and load-bearing: it leaves each cell's lifetime and momentum alone, so a freeze stops
        // a world without also resetting everything that was moving through it.
        let chilled = JS.toFloat32(-200)
        for i in 0 ..< cellCount {
            let cellType = type[i]
            // Air has no temperature worth setting, and bedrock is the world's container.
            if cellType == Element.empty || cellType == Element.bedrock { continue }
            temperature[i] = chilled
            // Separate tests rather than a chain, matching the original. Both read the element
            // as it was before either could have changed it, so nothing is converted twice.
            if cellType == Element.water || cellType == Element.acid
                || cellType == Element.oil || cellType == Element.saltWater
            {
                type[i] = Element.ice
            }
            if cellType == Element.lava {
                type[i] = Element.stone
            }
        }

        return PowderEventStart(shake: 0, sound: nil, soundIntensity: 1, followUp: nil)
    }
}

File: web/src/sim/powder-events.ts
/**
 * The four set-piece events: a meteor, a blast, a surge of water, and a deep freeze.
 *
 * ## Why these are here and not in the view
 *
 * They were written inline in `powder-view.tsx`, mixed in with sound calls and screen-shake
 * state. That put a set of world-mutating rules — radii, temperatures, which element replaces
 * which — somewhere no test can reach them, and it is the only reason they were never checked
 * against the native port the way everything else has been.
 *
 * Moving them here changes nothing about what they do. It makes them comparable.
 *
 * ## Sound and shake are described, not performed
 *
 * An event wants to shake the screen and make a noise, and neither belongs to the simulation.
 * So each function *reports* what should accompany the world change and leaves the caller to do
 * it — the same split the engine already uses for explosions via its `onBurst` callback.
 *
 * ## The delayed second half
 *
 * A meteor lands and then detonates a moment later; a blast goes off and is followed by two
 * flanking ones. That delay is what makes them read as events rather than as instant state
 * changes, so the timing is part of the definition and is returned rather than hidden in a
 * `setTimeout` inside a component.
 */

import type { PowderEngine } from "./powder-engine";
import { triggerExplosion } from "./powder/explosion";

export type PowderEventId = "meteor" | "blast" | "surge" | "freeze";

export type PowderEventSound = "meteor" | "explosion";

/** One explosion, as an event schedules it. */
export type PowderEventBlast = {
  x: number;
  y: number;
  radius: number;
  force: number;
  heat: number;
};

/** The second half of an event, to be run after `delayMs`. */
export type PowderEventFollowUp = {
  delayMs: number;
  blasts: PowderEventBlast[];
  shake: number;
  sound: PowderEventSound | null;
  soundIntensity: number;
};

/** What happened immediately, and what is still to come. */
export type PowderEventStart = {
  shake: number;
  sound: PowderEventSound | null;
  soundIntensity: number;
  followUp: PowderEventFollowUp | null;
};

const LAVA = 6;
const FIRE = 4;
const WATER = 2;
const STONE = 7;
const ACID = 8;
const OIL = 9;
const ICE = 13;
const SALT_WATER = 27;
const BEDROCK = 29;
const EMPTY = 0;

/**
 * Runs the immediate half of an event.
 *
 * @returns what should accompany it, including the delayed half if it has one.
 */
export function startPowderEvent(e: PowderEngine, id: PowderEventId): PowderEventStart {
  switch (id) {
    case "meteor":
      return meteor(e);
    case "blast":
      return blast(e);
    case "surge":
      return surge(e);
    case "freeze":
      return freeze(e);
  }
}

/** Runs the delayed half of an event. */
export function finishPowderEvent(e: PowderEngine, followUp: PowderEventFollowUp) {
  for (const b of followUp.blasts) {
    triggerExplosion(e, b.x, b.y, b.radius, b.force, b.heat);
  }
}

/**
 * A ball of lava and fire dropped from above, which detonates when it lands.
 *
 * The explosion is placed at a fixed fraction of the world's height rather than wherever the
 * material actually got to, because the material may not get anywhere — it can land on a
 * ceiling someone built. Deciding the impact point up front means the event always reads the
 * same.
 */
function meteor(e: PowderEngine): PowderEventStart {
  const cx = Math.floor(e.width / 2);
  const radius = 10;
  for (let dy = -radius; dy <= radius; dy++) {
    for (let dx = -radius; dx <= radius; dx++) {
      // A disc rather than a square, compared without a square root.
      if (dx * dx + dy * dy > radius * radius) continue;
      const x = cx + dx;
      const y = 14 + dy;
      // Checked before the draw below, so that a meteor near the edge of a narrow world
      // consumes exactly as many random numbers as there are cells it can actually fill.
      if (!e.isValid(x, y)) continue;
      e.setElementAt(x, y, Math.random() < 0.8 ? LAVA : FIRE, 2800);
      e.gridVy[e.getIndex(x, y)] = 18;
    }
  }
  return {
    shake: 16,
    sound: "meteor",
    soundIntensity: 1,
    followUp: {
      delayMs: 180,
      blasts: [{ x: cx, y: Math.floor(e.height * 0.65), radius: 28, force: 18, heat: 3000 }],
      shake: 22,
      sound: "explosion",
      soundIntensity: 2,
    },
  };
}

/** One large explosion at the centre, then two flanking it. */
function blast(e: PowderEngine): PowderEventStart {
  const cx = Math.floor(e.width / 2);
  const cy = Math.floor(e.height / 2);
  triggerExplosion(e, cx, cy, 36, 22, 3500);
  return {
    shake: 24,
    sound: "explosion",
    soundIntensity: 3,
    followUp: {
      delayMs: 120,
      blasts: [
        { x: cx - 22, y: cy - 14, radius: 22, force: 16, heat: 2800 },
        { x: cx + 22, y: cy + 14, radius: 22, force: 16, heat: 2800 },
      ],
      shake: 16,
      sound: null,
      soundIntensity: 1,
    },
  };
}

/** A wall of water down the left side, already moving right. */
function surge(e: PowderEngine): PowderEventStart {
  const startY = Math.floor(e.height * 0.28);
  const endX = Math.min(28, e.width - 4);
  for (let y = startY; y < e.height - 2; y++) {
    for (let x = 2; x < endX; x++) {
      e.setElementAt(x, y, WATER, 12);
      e.gridVx[e.getIndex(x, y)] = 14;
    }
  }
  return { shake: 0, sound: null, soundIntensity: 1, followUp: null };
}

/**
 * Chills everything at once: liquids become ice and lava sets to stone.
 *
 * Writes the grid directly rather than going through `setElementAt`, which is deliberate and
 * load-bearing — it leaves each cell's lifetime and momentum alone, so a freeze stops a world
 * without also resetting everything that was moving through it.
 */
function freeze(e: PowderEngine): PowderEventStart {
  for (let i = 0; i < e.gridType.length; i++) {
    const t = e.gridType[i];
    // Air has no temperature worth setting, and bedrock is the world's container.
    if (t === EMPTY || t === BEDROCK) continue;
    e.gridTemp[i] = -200;
    if (t === WATER || t === ACID || t === OIL || t === SALT_WATER) e.gridType[i] = ICE;
    if (t === LAVA) e.gridType[i] = STONE;
  }
  return { shake: 0, sound: null, soundIntensity: 1, followUp: null };
}
